src/name_gen.rs

```diff
@@ -1,15 +1,115 @@
+use std::collections::HashMap;
+use std::fs::OpenOptions;
+use std::io::{BufReader, BufRead, Write};
+
+use rand::thread_rng;
+use rand::distributions::{Weighted, WeightedChoice, Sample};
+use epic;
+use fs2::FileExt;
+
+
+/// Generates a random message, selecting an epic and a message format.
+pub fn random_msg() -> String {
+    let epic = select_epic();
+    epic_msg(epic)
+}
+
 pub fn super_mario_epic(epic_title: &str) -> String {
     format!("Super Mario {}", epic_title)
 }
 
+/// This selects an epic, preferring ones that haven't been chosen too recently.
+/// It works by using a file "epics.log" listed in order as they're read. It
+/// then uses the order to prioritize the random selection, and writes its
+/// choice back into the file.
+fn select_epic() -> &'static epic::Epic {
+    let num_skip = 5;
+    let num_ease = 10;
+    let num_weight = num_skip + num_ease;
+
+    // @Robustness: Handle these unwraps better?
+    let file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open("epics.log")
+        .unwrap();
+    file.lock_exclusive().unwrap();
+
+    // Make the weights, set 0 for the 5 most recent...
+    let mut weights = HashMap::new();
+    let mut file = BufReader::new(file);
+    let mut lines = Vec::new();
+    let mut line = String::new();
+    while let Ok(x) = file.read_line(&mut line) {
+        if x == 0 {
+            break;
+        }
+        lines.push(String::from(line.trim()));
+        line.clear();
+    }
+
+    for line in lines.iter().rev().take(num_skip) {
+        weights.entry(line.clone()).or_insert(0);
+    }
+    // Then doubling weight for each of the next 10 items
+    let mut weight = 1;
+    for line in lines.iter().rev().skip(num_skip).take(num_ease) {
+        weights.entry(line.clone()).or_insert(weight);
+        weight *= 2;
+    }
+    // Any item that remains gets a weight of 1024
+
+    let mut epics: Vec<_> = epic::EPICS
+        .iter()
+        .map(|epic| {
+            Weighted {
+                weight: *weights.get(epic.title).unwrap_or(&weight),
+                item: epic,
+            }
+        })
+        .collect();
+    let choice = WeightedChoice::new(&mut epics[..]).sample(&mut thread_rng());
+
+    // We record which one we chose so it doesn't get chosen too soon.
+    lines.push(choice.title.into());
+    let mut file = file.into_inner();
+    // @Robustness: Handle these unwraps?
+    // We only write the weighted entries so that it doesn't grow forever
```

Comment:
Hmm, running the program about 100 times results in a log file that definitely doesn't stop growing. I think the problem is that `lines.len()` (measured in how many entries the Vector has) isn't the same thing as the length of the file (measured in bytes?).

Reply:
On fish I get this:
```
$ for i in (seq 32)
    target/debug/epic_mario > /dev/null
    wc -l epics.log
end
       1 epics.log
       2 epics.log
       3 epics.log
       4 epics.log
       5 epics.log
       6 epics.log
       7 epics.log
       8 epics.log
       9 epics.log
      10 epics.log
      11 epics.log
      12 epics.log
      13 epics.log
      14 epics.log
      15 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
      16 epics.log
```
which looks like it's not growing to me

Reply:
Hmm. Seems like I was mistaken then. I'll approve changes now.